botorch/acquisition/knowledge_gradient.py

```diff
@@ -46,12 +47,456 @@
+        Parameters
+        ----------
+        a
+            Vector of intercepts describing a set of straight lines
+        b
+            Vector of slopes describing a set of straight lines
+        Returns
+        -------
+        KGCB
+            average height of the epigraph
+        """
+
+        a = a.squeeze()
+        b = b.squeeze()
+        assert len(a) > 0, "must provide slopes"
+        assert len(a) == len(b), f"#intercepts != #slopes, {len(a)}, {len(b)}"
+
+        maxa = torch.max(a)
+
+        if torch.all(torch.abs(b) < 0.000000001):
+            return torch.Tensor([0])  # , np.zeros(a.shape), np.zeros(b.shape)
+
+        # Order by ascending b and descending a. There should be an easier way to do this
+        # but it seems that pytorch sorts everything as a 1D Tensor
+
+        ab_tensor = torch.vstack([-a, b]).T
+        ab_tensor_sort_a = ab_tensor[ab_tensor[:, 0].sort()[1]]
+        ab_tensor_sort_b = ab_tensor_sort_a[ab_tensor_sort_a[:, 1].sort()[1]]
+        a = -ab_tensor_sort_b[:, 0]
+        b = ab_tensor_sort_b[:, 1]
+
+        # exclude duplicated b (or super duper similar b)
+        threshold = (b[-1] - b[0]) * 0.00001
+        diff_b = b[1:] - b[:-1]
+        keep = diff_b > threshold
+        keep = torch.cat([torch.Tensor([True]), keep])
+        keep[torch.argmax(a)] = True
+        keep = keep.bool()  # making sure 0 1's are transformed to booleans
+
+        a = a[keep]
+        b = b[keep]
+
+        # initialize
+        idz = [0]
+        i_last = 0
+        x = [-torch.inf]
+
+        n_lines = len(a)
+        # main loop TODO describe logic
+        # TODO not pruning properly, e.g. a=[0,1,2], b=[-1,0,1]
+        # returns x=[-inf, -1, -1, inf], shouldn't affect kgcb
+        while i_last < n_lines - 1:
+            i_mask = torch.arange(i_last + 1, n_lines)
+            x_mask = -(a[i_last] - a[i_mask]) / (b[i_last] - b[i_mask])
+
+            best_pos = torch.argmin(x_mask)
+            idz.append(i_mask[best_pos])
+            x.append(x_mask[best_pos])
+
+            i_last = idz[-1]
+
+        x.append(torch.inf)
+
+        x = torch.Tensor(x)
+        idz = torch.LongTensor(idz)
+        # found the epigraph, now compute the expectation
+        a = a[idz]
+        b = b[idz]
+
+        normal = Normal(torch.zeros_like(x), torch.ones_like(x))
+
+        pdf = torch.exp(normal.log_prob(x))
+        cdf = normal.cdf(x)
+
+        kg = torch.sum(a * (cdf[1:] - cdf[:-1]) + b * (pdf[:-1] - pdf[1:]))
+        kg -= maxa
+        return kg
+
+    @staticmethod
+    def construct_z_vals(nz: int, device: Optional[torch.device] = None) -> Tensor:
+        """make nz equally quantile-spaced z values"""
+
+        quantiles_z = (torch.arange(nz) + 0.5) * (1 / nz)
+        normal = torch.distributions.Normal(0, 1)
+        z_vals = normal.icdf(quantiles_z)
+        return z_vals.to(device=device)
+
+
+class ContinuousKnowledgeGradient(MCAcquisitionFunction):
```

Comment:
can we merge common code in continuousKG and hybrid KG? Make the continuos KG class first, then hybridKG inherits from it so that the MC-kg code is common to both methods?